expand/param.go

```diff
@@ -78,31 +78,72 @@ func (cfg *Config) paramExp(pe *syntax.ParamExp) (string, error) {
 		}
 	}
 
-	str, err := cfg.varInd(vr, index)
-	if err != nil {
-		return "", err
-	}
-	slicePos := func(n int) int {
-		if n < 0 {
-			n = len(str) + n
-			if n < 0 {
-				n = len(str)
+	var sliceOffset, sliceLen int
+	if s := pe.Slice; s != nil {
+		var err error
+		if s.Offset != nil {
+			sliceOffset, err = Arithm(cfg, s.Offset)
+			if err != nil {
+				return "", err
+			}
+		}
+		if s.Length != nil {
+			sliceLen, err = Arithm(cfg, s.Length)
+			if err != nil {
+				return "", err
 			}
-		} else if n > len(str) {
-			n = len(str)
 		}
-		return n
 	}
-	elems := []string{str}
+
+	var (
+		str   string
+		elems []string
+
+		splatOp bool // true if var has been accessed with * or @ index
+		varInd  = true
+	)
+
 	switch nodeLit(index) {
 	case "@", "*":
 		switch vr.Kind {
 		case Unset:
 			elems = nil
+			splatOp = true
 		case Indexed:
+			splatOp = true
+			varInd = false
 			elems = vr.List
+			indexPos := func(n int) int {
```

Comment:
perhaps call it `slicePos` too? that's still what it's doing - it's just using `len` of the `elems` slice, as opposed to a string.